Clean up startup logging, fixate with test.

jakewins opened the pull request:
- Remove Logback Slf4j bridge from classpath, meaning slf4j pipes to j.u.l,
  meaning it gets piped into our log framework.
- Turn most startup log statements into debug level statements
- Make actual info-level startup statements conistent and friendlier.


## Before

    16:09:08.968 [main] DEBUG i.n.u.i.l.InternalLoggerFactory - Using SLF4J as the default logging framework
    2015-10-27 16:09:12.481-0500 INFO  Successfully started database
    2015-10-27 16:09:12.502-0500 INFO  Starting HTTP on port 7474 (8 threads available)
    2015-10-27 16:09:12.704-0500 INFO  Enabling HTTPS on port 7473
    2015-10-27 16:09:12.765-0500 INFO  Mounting static content at /webadmin
    2015-10-27 16:09:12.800-0500 INFO  Mounting static content at /browser
    2015-10-27 16:09:13.655-0500 INFO  Remote interface ready and available at http://localhost:7474/
    2015-10-27 16:17:47.162-0500 INFO  Neo4j Server shutdown initiated by request
    2015-10-27 16:17:47.207-0500 INFO  Successfully stopped database
    2015-10-27 16:17:47.207-0500 INFO  Successfully shutdown database

## After

    2015-10-27 19:07:55.303-0500 INFO  Neo4j is starting..
    2015-10-27 19:07:59.018-0500 INFO  Neo4j has started! The HTTP API and the Neo4j Browser are now available at http://localhost:7474/
    2015-10-27 19:07:59.019-0500 INFO  Neo4j is stopping..
    2015-10-27 19:07:59.063-0500 INFO  Neo4j has stopped, see you next time!

These may not be exactly what we want, but this PR at least introduces a test that fixates something sensible, that we can then easily iterate on. It also resolves the issue of slf4j piping into logback.

MishaDemianenko commented on community/server/src/test/java/org/neo4j/server/integration/StartupLoggingIT.java:
```diff
@@ -0,0 +1,124 @@
+
+        public List<String> lines() throws UnsupportedEncodingException
+        {
+            String s = this.captured.toString( "UTF-8" );
```
No more encoding names and `UnsupportedEncodingException`s please : ), 
use `UTF8` helper class or `StandardCharsets` when suitable

jakewins replied:
I would, but toString on `ByteArrayOutputStream` only takes a string charset :( I don't think it's a huge issue here, since this is in a test, so there's no catching it anywhere anyway.

I did notice that i missed a system.out on the next line tho..

MishaDemianenko replied:
StandardCharsets.UTF8.name() if possible then, first step of cleaning up encoding names usages are already in (3.0 only, sorry), will try to keep people away from that way.
Yep it's test, till somebody will copy paste it elsewhere :)

MishaDemianenko commented on community/server/src/test/java/org/neo4j/server/integration/StartupLoggingIT.java:
```diff
@@ -0,0 +1,124 @@
+    public void shouldLogHelpfulStartupMessages() throws Throwable
+    {
+        // Given
+        CaptureStdOut stdout = new CaptureStdOut();
```
`SuppressOutput` should help here

jakewins commented:
@MishaDemianenko fair enough :) Swapped to using `StandardCharset` and `SuppressOutput`

MishaDemianenko commented:
@jakewins thank you :bow:

cleishm commented on community/server/src/main/java/org/neo4j/server/AbstractNeoServer.java:
```diff
@@ -361,7 +363,7 @@ private void configureWebServer() throws Exception
         if ( sslEnabled )
         {
-            log.info( "Enabling HTTPS on port %s", sslPort );
+            log.debug( "Enabling HTTPS on port %s", sslPort );
```
All these changes to the actual output should probably go in 3.0

cleishm commented on community/server/src/test/java/org/neo4j/server/integration/StartupLoggingIT.java:
```diff
@@ -0,0 +1,100 @@
+        List<String> captured = suppressed.getOutputVoice().lines();
+        assertThat( captured, matchesLines(
+                info( "Neo4j is starting.." ),
+                info( "Neo4j has started! The HTTP API and the Neo4j Browser are now available at http://.*:7474/" ),
```
We shouldn't change this output in 2.3, as people do log scraping for existing values.

cleishm commented on community/server/pom.xml:
```diff
@@ -159,7 +159,7 @@
     <dependency>
       <groupId>ch.qos.logback</groupId>
-      <artifactId>logback-classic</artifactId>
+      <artifactId>logback-core</artifactId>
```
Doesn't this have the effect of changing dependencies in a point release?

tinwelint replied:
Still find it odd that we have logback as dependency in 2.3

jakewins replied:
@tinwelint It's because of HTTP logging in Jetty :/. I think we'd be able to write our own thing for that as well, if we wanted to get rid of logback altogether. I def. understand the reasons for wanting to not expose logback config to users, but I still haven't understood why we want to remove logback itself.

@cleishm yep

tinwelint replied:
@jakewins Well, for one thing we should make up our minds, either we use logback or we don't. As it is now it's semi-removed... which is just confusing. Should be dead easy to write our own adapter thing there, right?

cleishm replied:
@tinwelint / @benbc We actually do have our own adapter thing for netty, called `Netty4LoggerFactory`, and [it is set here](https://github.com/neo4j/neo4j/blob/3.0/community/server/src/main/java/org/neo4j/server/Bootstrapper.java#L84). Netty _only uses logback for a single message on initialization_, due to an [odd behaviour in the static intializer of the netty `InternalLoggerFactory`](https://github.com/netty/netty/blob/master/common/src/main/java/io/netty/util/internal/logging/InternalLoggerFactory.java#L37-L66) that was added in 4.0. I consider that a bug in netty, and will probably raise it to the netty team.

Given it's only the one use of logback in the static initializer, we could just do the very simple thing in 2.3 of silencing that by doing something like this:

```
        final PrintStream originalStdout = System.out;
        System.setOut( NullPrintStream.getInstance() );
        InternalLoggerFactory.setDefaultFactory( new Netty4LoggerFactory( userLogProvider ) );
        System.setOut(originalStdout);
```

If you'd like to try that approach, I have a commit that does exactly that here: https://github.com/cleishm/neo4j/commit/95f99eaf992b9bdcbd547af773f7b8ba606f544f

jakewins replied:
@cleishm Netty does not use logback, it uses Slf4j. Debug logging to Slf4j is not a bug, the problem is us having an Slf4j handler on the classpath (eg. logback-classic) without having configured it. In fact, Netty sends the exact same log message to debug with the fix in this PR, the difference being that j.u.l is piped to be handled by us.

@tinwelint I've still not understood the reasoning for us building our own logging system - I understand us having our own logging _interfaces_, but I don't understand why we are building our own logging backend. Eg. I wrote that Netty4LoggerFactory that CL mentioned, and it's super frustrating, because we are going to have to do that to every single third party library we pull in, and we are going to keep forgetting and keep having to figure out how each new library does custom logging. 

If we used Logback and configured it properly, it registers itself as a standard handler for Slf4j, and logging would "just work" for any normal java lib we pull in. Alternatively we could build our own Slf4j integration - but again _why_ are we doing this when there are really performant alternatives out there already, which we have on the classpath anyway?

Replacing the HTTP logging, which is on the hot path, would mean we also need to match Logbacks performance - keeping in mind that it uses disruptor under the hood to offload logging work. Eg; that's a lot of engineering to do.

Is there an upside to building this in-house that I'm not seeing?
</rant>

chrisvest commented on community/server/src/test/java/org/neo4j/server/integration/StartupLoggingIT.java:
```diff
@@ -0,0 +1,100 @@
+package org.neo4j.server.integration;
+
```
Needs a juicy little license header.

benbc commented:
I think this is a great improvement, but I'm afraid it'll have to wait to 3.0.

I'd like to do a bigger review of logging for 3.0, so don't feel you need to struggle to consensus on this now if you don't have the energy for it.

jakewins committed:
Temporary workaround to avoid debug logging from Netty.

- Because this will go into patch release, silence stdout while Netty
  logging loads. In next major or minor release, we'll fix this by
  either removing Slf4j from the classpath or configuring it properly.

jakewins commented:
@benbc @cleishm removed the actual logging changes, so now it just makes sure Slf4j goes into j.u.l, which goes into our logging, plus a test to fixate it.

jakewins commented:
@cleishm So, thinking this through, the two options we know of are;

```
// A
disable stdout
load logback+slf4j
enable stdout

// B
remove slf4j from classpath
```

I can't say I'm particularly comfortable with either - but in the scale of things, the `logback-classic` dep may contain things that break other behaviors if we remove them, so I think I agree that the `disable stdout/load/enable stdout` approach is the least risky option.

Which is to say: How about you take the test and associated fix for `SuppressOutput` from this PR, and open a new PR with the stdout approach, and then we can revert that in 3.0 and fix Slf4j properly there?

cleishm commented:
Agreed that the disable/enable stdout is a hack, but probably the least invasive. In 3.0 we should do both - disable stdout, and setup slf4j/logback/whatever properly.

I'll try to find some time this weekend to make a PR as you've suggested.

On your comment earlier: there's no reason not to use a 3rd party log framework in neo4j server. The logging in kernel doesn't use one, simply to avoid conflicting dependencies. It provides a way to inject in any logger you like, and it'll use it for user logging. The in-house logging code is only used for the internal diagnostics log (messages.log) since it has very simple needs, shouldn't be managed by the user code (in embedded), and having it in our code avoids the dependency conflict issue. None of that is a problem in server, so we could certainly use a 3rd party approach - just as long as it gives nice, clean output. That said, the needs of the user log are likely simple enough there also that we can just use the in-house code there also (at least until we start talking about remote logging, etc). For HTTP logging, I don't know. As you mention, it's probably hot enough that we should use a fancy 3rd party one that does disrupter-magicy-performance-stuff :)

benbc commented:
@cleishm said:

> The in-house logging code is only used for the internal diagnostics log (messages.log) since it has very simple needs[.] [...] For HTTP logging [...] it's probably hot enough that we should use a fancy 3rd party one that does disrupter-magicy-performance-stuff

We recently had a massive performance bug caused by logging to `messages.log` on the hot path in the Kernel. Perhaps we should consider fancy logging there as well.

tinwelint commented:
@jakewins license headers

jakewins commented:
@cleishm I bumped this PR to use your stdout workaround instead of the logback change.

@tinwelint do you mind taking this PR in your strong arms?

tinwelint commented:
@jakewins not at all
